Add BinPlaceConfiguration for test projects to have a working directory per BuildConfiguration

safern opened the pull request:
This changes depends on: https://github.com/dotnet/buildtools/pull/1652 and a BuildTools update once that is merged. 

Fixes: https://github.com/dotnet/buildtools/issues/1587

## New Behavior with this change

With this change we will have the next behavior when building/running tests:

- When building the tests we will still bin place the test assets to the `OutputDirectory` (`corefx\bin\OSGroup.Configuration\TestProject\TargetGroup/`)
- When running the tests we will bin place the tests assets to this new [`TestPath`](https://github.com/dotnet/corefx/pull/23633/files#diff-4de71e7a0fd952be821f8af2768281dcR34) which will be the `TestPath`. This new path will be the test's working directory where we will bin place the RunTests.cmd and necessary dependencies in tests.targets. Also the `testResults.xml` will be found here after the test execution.
- `TestPath` is set to be `bin/tests/<TestProject>/<BuildConfiguration>/` (i.e `corefx\bin\tests\System.Collections.Immutable.Tests\netcoreapp-Windows_NT-Debug-x64`) in `dir.props`
- This `TestPath` will be added to the `BinPlaceDir` in `FrameworkTargeting.targets` when `BinPlaceTest` is set to true. 
- Then in the [`BinPlaceFiles`](https://github.com/safern/buildtools/blob/4d39247ff2c152ac50ade6128a44afc6f9993ca2/src/Microsoft.DotNet.Build.Tasks/PackageFiles/FrameworkTargeting.targets#L243) target the test assets will be hardlinked to the `TestPath`.

This will isolate every test execution per build configuration and will allow people to run the same test project for all of the supported configurations on the same repo without having to clean the repo.

cc: @weshaggard @ericstj @joperezr 

FYI: @danmosemsft

ericstj commented on dir.targets:
```diff
@@ -87,6 +88,13 @@
   </ItemGroup>
 
+  <!-- For tests we want to separate the bin directory and the working directory to have a working directory per build configuration. -->
+  <ItemGroup Condition="'@(BinPlaceConfiguration)' == '' AND '$(BinPlaceTest)' == 'true'">
```
There is no need to seperate this out.  You're already applying the BinPlaceTest condition on the TestWorkingPath in the binplace targets so this is redundant.

safern replied:
Indeed. Thanks.

ericstj replied:
Also please consider reusing the $(TestPath) property to ensure that these two paths don't get out of sync.

ericstj commented on dir.targets:
```diff
@@ -22,9 +22,10 @@
+  <ItemGroup Condition="'@(BinPlaceConfiguration)' == '' AND '$(IsTestProject)' != 'true'">
     <!-- binplace to directories for the target vertical -->
     <BinPlaceConfiguration Include="$(_bc_TargetGroup)-$(_bc_OSGroup)">
       <RefPath>$(BuildConfigurationRefPath)</RefPath>
```
Just add the test  path here.

safern replied:
Yes you're right.

safern commented:
I've addressed PR Feedback and tested locally that it works correctly with the following scenarios:
- run a test project that has only netstandard build configuration in configuration.props for both uap and netcoreapp and validated that it created the correct folders and that they have the correct stuff in there.
- run a test project that has uap build configuration for both uap and uapaot and validated correctness, works fine as well.

Will update the details in buildtools PR to add the descriptions of the new behavior.

ericstj commented:
> TestPath is set to be bin/tests/<TestProject>/<BuildConfiguration>/ (i.e corefx\bin\tests\System.Collections.Immutable.Tests)

Missing BuildConfiguration in the example.

Have you examined all the targets to make sure everything that is BuildConfiguration-specifc operates on TestPath  and not the intermediate or output path of the test project?

safern commented:
> Approved assuming you've examined all cases where test targets might use intermediate or output to ensure they correctly use your new path.

It uses `TestPath` which I'm setting to the same path in `dir.props` on this PR. If `TestPath` is empty we will set it to the default which is the `OutputDirectory`.

I just scanned tests.targets for output and we put everything in the test path specific for intermediate stuff (dependency files, configuration files to produce pri file, etc).

joperezr commented on src/System.Data.DataSetExtensions/tests/Mono/DataTableExtensionsTest.cs:
```diff
@@ -40,6 +40,8 @@ namespace MonoTests.System.Data
 {
     public class DataTableExtensionsTest
     {
+        private string _testDataSet = "testdataset1.xml";
```
Why do we need these changes?

safern replied:
Right now the way that we bin place and run the tests is through CopyToOutputDirectory and through the build output to the OutputDirectory, the way that the build bin places items is with the same path with the same structure as we have it in the test project. (i.e if we have a txt file under `test-project-folder/myFiles/text.txt`it will be bin placed under a myFiles directory in the output directory). 

With this new changes, we would bin place the working directory through the bin place targets and this bin places all the dependencies under the same folder without subdirectories, so I needed to update the tests that reference files to an specific path under the working directory, to reference this files to the root working directory.

safern commented:
test UWP CoreCLR x64 Debug Build please
test Linux x64 Release Build please

safern commented:
@danmosemsft seems like a flaky test in System.Drawing that is crashing on Linux. Safe to merge?

danmosemsft commented:
@safern yes thanks for asking. @mellinoe fyi.

safern commented:
> @safern yes thanks for asking.

Ok will merge then. Thanks.

safern commented:
FYI: @davidsh for the networking test changes to match this new bin placing configuration.
